src/FsAutoComplete/LspServers/ProjectWorkspace.fs

```diff
@@ -0,0 +1,298 @@
+    }
+
+  let rec private createReferences
+    (cachedSnapshots)
+    (inMemorySourceFiles: amap<string<LocalPath>, aval<VolatileFile>>)
+    (sourceTextFactory: aval<ISourceTextFactory>)
+    (loadedProjectsA: amap<string<LocalPath>, ProjectOptions>)
+    (p: ProjectOptions)
+    =
+    logger.debug (
+      Log.setMessage "Creating references for {projectFileName}"
+      >> Log.addContextDestructured "projectFileName" p.ProjectFileName
+    )
+
+    loadedProjectsA
+    |> AMap.filter (fun k _ ->
+      p.ReferencedProjects
+      |> List.exists (fun x -> normalizePath x.ProjectFileName = k))
+    |> AMap.map (fun _ proj ->
+      if proj.ProjectFileName.EndsWith ".fsproj" then
+
+        let resolvedTargetPath =
+          aval {
+            // TODO: Find if this needs to be adaptive, unsure if we need to check if the file has changed on disk if we need a new snapshot
+            let! _ = AdaptiveFile.GetLastWriteTimeUtc proj.ResolvedTargetPath
+            return proj.ResolvedTargetPath
+          }
+
+        proj
+        |> optionsToSnapshot
+          cachedSnapshots
+          inMemorySourceFiles
+          sourceTextFactory
+          (createReferences cachedSnapshots inMemorySourceFiles sourceTextFactory loadedProjectsA)
+        |> createReferencedProjectsFSharpReference resolvedTargetPath
+      else
+        // TODO: Find if this needs to be adaptive or if `getStamp` in a PEReference will be enough to break thru the caching in FCS
+        loadFromDotnetDll proj |> AVal.constant)
+    |> AMap.toASetValues
+
+  and optionsToSnapshot
+    (cachedSnapshots: Dictionary<_, _>)
+    (inMemorySourceFiles: amap<_, aval<VolatileFile>>)
+    (sourceTextFactory: aval<ISourceTextFactory>)
+    (mapReferences: ProjectOptions -> aset<aval<FSharpReferencedProjectSnapshot>>)
+    (p: ProjectOptions)
+    =
+    let normPath = Utils.normalizePath p.ProjectFileName
+
+    match cachedSnapshots.TryGetValue normPath with
+    | true, snapshot ->
+      logger.debug (
+        Log.setMessage "optionsToSnapshot - Cache hit - {projectFileName}"
+        >> Log.addContextDestructured "projectFileName" p.ProjectFileName
+      )
+
+      snapshot
+    | _ ->
+      logger.debug (
+        Log.setMessage "optionsToSnapshot - Cache miss - {projectFileName}"
+        >> Log.addContextDestructured "projectFileName" p.ProjectFileName
+      )
+
+      let projectName = AVal.constant p.ProjectFileName
+      let projectId = AVal.constant p.ProjectId
+
+
+      let sourceFiles = // alist because order matters for the F# Compiler
+        p.SourceFiles
+        |> AList.ofList
+        |> AList.map Utils.normalizePath
+        |> AList.map (fun sourcePath ->
+
+          aval {
+            match! inMemorySourceFiles |> AMap.tryFind sourcePath with
+            | Some volatileFile -> return! volatileFile |> AVal.map createFSharpFileSnapshotInMemory
+            | None -> return! createFSharpFileSnapshotOnDisk sourceTextFactory sourcePath
```

Comment:
Changes to source files, either in memory or on disk will cause snapshots to be recreated.